torch/csrc/autograd/input_buffer.cpp

```diff
@@ -65,20 +93,46 @@ namespace torch { namespace autograd {
                         && device_of(var) == opt_consumer_stream->device();
     if (on_producer && on_consumer) {
       // (2) CUDA variable with producer and consumer sharing a device
-      //     Accumulation happens on consumer's stream
       opt_accumulate_stream = opt_consumer_stream;
-      if (opt_producer_stream != opt_consumer_stream) {
-        // (2a) Syncs consumer with producer
+      if (opt_accumulate_stream != opt_producer_stream) {
+        // (2a) Sync accumulate with producer
         auto event = c10::Event{c10::DeviceType::CUDA};
         event.record(*opt_producer_stream);
-        opt_consumer_stream->wait(event);
+        opt_accumulate_stream->wait(event);
       }
     } else {
-      // (3) CUDA variable with multiple devices
-      //     Accumulation happens on variable's device's default stream
+      // hack to get var's device's default stream that will compile even for CPU-only builds
       const auto guard = c10::impl::VirtualGuardImpl{c10::DeviceType::CUDA};
-      const auto default_stream = guard.getDefaultStream(*device_of(var));
-      opt_accumulate_stream = default_stream;
+      const auto var_device_default_stream = guard.getDefaultStream(*device_of(var));
```

Comment:
In this section, can we logically split the code to determine which stream to accumulate on, then have an optional of a "sync stream" which, if populated, the accumulate stream syncs with (in one place?)